File: common/src/main/java/dev/mayaqq/estrogen/registry/items/DreamBottleItem.java
package dev.mayaqq.estrogen.registry.items;

import dev.mayaqq.estrogen.registry.EstrogenBlocks;
import dev.mayaqq.estrogen.registry.EstrogenSounds;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.item.ItemNameBlockItem;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;

public class DreamBottleItem extends ItemNameBlockItem {
    public DreamBottleItem(Properties properties) {
        super(EstrogenBlocks.DREAM_BLOCK.get(), properties);
    }

    @Override
    public SoundEvent getPlaceSound(BlockState state) {
        return EstrogenSounds.DREAM_BLOCK_PLACE.get();
    }

    @Override
    public InteractionResult place(BlockPlaceContext context) {
        InteractionResult result = super.place(context);
        if (result == InteractionResult.SUCCESS) {
            context.getPlayer().getInventory().placeItemBackInInventory(Items.GLASS_BOTTLE.getDefaultInstance());
        }
        return result;
    }

    @Override
    public InteractionResult useOn(UseOnContext context) {
        Level level = context.getLevel();
        InteractionResult actionResult = super.useOn(context);
        if (!level.isClientSide && actionResult.consumesAction() && context.getPlayer() != null && !context.getPlayer().isCreative()) {
            context.getPlayer().getInventory().placeItemBackInInventory(Items.GLASS_BOTTLE.getDefaultInstance());
        }
        return actionResult;
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/content/items/DreamBottleItem.kt
package dev.mayaqq.estrogen.content.items

import dev.mayaqq.cynosure.utils.toVector3d
import dev.mayaqq.estrogen.content.EstrogenBlocks
import dev.mayaqq.estrogen.content.EstrogenSounds
import dev.mayaqq.estrogen.content.blockEntities.DreamBlockEntity
import dev.mayaqq.estrogen.content.blocks.DreamBlock
import net.minecraft.sounds.SoundEvent
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemNameBlockItem
import net.minecraft.world.item.Items
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.item.context.UseOnContext
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.shapes.BooleanOp
import net.minecraft.world.phys.shapes.Shapes

class DreamBottleItem(p0: Block, p1: Properties) : ItemNameBlockItem(p0, p1) {

    public override fun getPlaceSound(state: BlockState): SoundEvent = EstrogenSounds.DREAM_BLOCK_PLACE

    override fun getPlacementState(context: BlockPlaceContext): BlockState? {
        return super.getPlacementState(context)?.setValue(DreamBlock.PERSISTENT, true)
    }

    override fun useOn(context: UseOnContext): InteractionResult {
        val level = context.level
        val actionResult = super.useOn(context)
        if (!level.isClientSide && actionResult.consumesAction() && context.player != null && !context.player!!.isCreative) {
            context.player!!.inventory.placeItemBackInInventory(Items.GLASS_BOTTLE.defaultInstance)
        }
        return actionResult
    }
}